### PR TITLE
server: fix prepared cursor select

### LINKED ISSUE #29570: jdbc test faild with `runtime error: index out of range`
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1.  git clone tidb-test
2. cd jdbc_test
2. start tidb, replace the host and port and run the following command.
```bash
 mvn "-Dcom.mysql.jdbc.testsuite.url=jdbc:mysql://$HOST:$PORT/test?user=root&useSSL=false&useServerPrepStmts=true" '-Dtest=CursorFetchTest' test
```

### 2. What did you expect to see? (Required)
execute successfully 
### 3. What did you see instead (Required)
```
RROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.475 s <<< FAILURE! - in pingcap.tidb.test.regression.CursorFetchTest
[ERROR] testBasicFetch(pingcap.tidb.test.regression.CursorFetchTest)  Time elapsed: 0.383 s  <<< ERROR!
java.sql.SQLException: runtime error: index out of range [0] with length 0
        at pingcap.tidb.test.regression.CursorFetchTest.testBasicFetch(CursorFetchTest.java:42)

[ERROR] testFetchPart(pingcap.tidb.test.regression.CursorFetchTest)  Time elapsed: 0.064 s  <<< ERROR!
java.sql.SQLException: runtime error: index out of range [0] with length 0
        at pingcap.tidb.test.regression.CursorFetchTest.testFetchPart(CursorFetchTest.java:67)

[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   CursorFetchTest.testBasicFetch:42 » SQL runtime error: index out of range [0] ...
[ERROR]   CursorFetchTest.testFetchPart:67 » SQL runtime error: index out of range [0] w...
[INFO]
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  6.309 s
[INFO] Finished at: 2021-11-08T18:15:53+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.0:test (default-test) on project tidb-jdbc-test: There are test failures.
[ERROR]
```
this maybe introduced by #28969, the test would be pass before this pr
### 4. What is your TiDB version? (Required)
```
╰─$ ./bin/tidb-server -V
Release Version: v5.4.0-alpha-40-gfdfba3e
Edition: Community
Git Commit Hash: fdfba3e4fafee8906d2dde6d737721608035aa6a
Git Branch: HEAD
UTC Build Time: 2021-11-08 10:14:53
GoVersion: go1.16.1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by xhebox**
Signed-off-by: xhe <xw897002528@gmail.com>

### What problem does this PR solve?

Issue Number: close #29570

Problem Summary: Stored chunk are reset by the allocator, leading to nil pointer panic, avoid the usage of the allocator. FYI, the previous PR, #29580.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
Refer scripts in #29570

### Release note

```release-note
None
```

### CODE REVIEW

**Review by tiancaiamao — APPROVED** [review]

File: server/conn.go [hunk 1]
```diff
@@ -2190,9 +2190,13 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
...
 	fetchedRows := rs.GetFetchedRows()
+	// if fetchedRows is not enough, getting data from recordSet.
 	for len(fetchedRows) < fetchSize {
-		// if fetchedRows is not enough, getting data from recordSet.
-		req := rs.NewChunk(cc.chunkAlloc)
+		// NOTE: chunk should not be allocated from the allocator
+		// the allocator will reset every statement
+		// but it maybe stored in the result set among statements
+		// ref https://github.com/pingcap/tidb/blob/7fc6ebbda4ddf84c0ba801ca7ebb636b934168cf/server/conn_stmt.go#L233-L239
+		req := rs.NewChunk(nil)
```
Comment (lcwangchao) [inline server/conn.go:2199]: Is it better to reuse the chunk and call `req.Reset()` for every loop for reduce allocs ? 

rest LGTM .

Comment (bb7133) [reply server/conn.go:2199]: @xhebox What do u think?

Comment (xhebox) [reply server/conn.go]: addressed.

**Review by lcwangchao — APPROVED** [review]

**Pull request merged**